### ISSUE TITLE
Add better output comparator for DSL plan testing

### ISSUE THREAD

**Issue by maciejgrzybek** [prestodb/presto#5967]
On DSL test failure it's hard to follow the output. Maybe some sort of highlighting would be helpful.

E.g. of the problematic output:

``` sql
java.lang.AssertionError: Plan does not match:
 - Output[sum_quantity_A, avg_discount_B, min_tax_A] => [expr_17:double, expr_18:double, expr_19:double]
        sum_quantity_A := expr_17
        avg_discount_B := expr_18
        min_tax_A := expr_19
    - Project => [expr_17:double, expr_18:double, expr_19:double]
            expr_17 := "expr"
            expr_18 := "expr_15"
            expr_19 := "expr_16"
        - Project => [expr:double, expr_15:double, expr_16:double]
                expr := "sum_12"
                expr_15 := "avg_13"
                expr_16 := "max_14"
            - Project => [sum_12:double, avg_13:double, max_14:double]
                    sum_12 := "sum_9"
                    avg_13 := "avg_10"
                    max_14 := "max_11"
                - Project => [sum_9:double, avg_10:double, max_11:double]
                        sum_9 := "sum"
                        avg_10 := "avg"
                        max_11 := "max"
                    - Window[partition by (orderkey_7), order by (tax_8)] => [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint, quantity_0:double, suppkey_1:bigint, orderkey_2:bigint, sum:double, discount_3:double, partkey_4:bigint, receiptdate_5:date, avg:double, extendedprice_6:double, orderkey_7:bigint, tax_8:double, max:double]
                            max := max("extendedprice_6")
                        - Project => [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint, quantity_0:double, suppkey_1:bigint, orderkey_2:bigint, sum:double, discount_3:double, partkey_4:bigint, receiptdate_5:date, avg:double, extendedprice_6:double, orderkey_7:bigint, tax_8:double]
                                extendedprice_6 := "extendedprice"
                                orderkey_7 := "orderkey_2"
                                tax_8 := "tax"
                            - Window[partition by (partkey_4), order by (receiptdate_5)] => [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint, quantity_0:double, suppkey_1:bigint, orderkey_2:bigint, sum:double, discount_3:double, partkey_4:bigint, receiptdate_5:date, avg:double]
                                    avg := avg("discount_3")
                                - Project => [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint, quantity_0:double, suppkey_1:bigint, orderkey_2:bigint, sum:double, discount_3:double, partkey_4:bigint, receiptdate_5:date]
                                        discount_3 := "discount"
                                        partkey_4 := "partkey"
                                        receiptdate_5 := "receiptdate"
                                    - Window[partition by (suppkey_1), order by (orderkey_2)] => [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint, quantity_0:double, suppkey_1:bigint, orderkey_2:bigint, sum:double]
                                            sum := sum("quantity_0")
                                        - Project => [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint, quantity_0:double, suppkey_1:bigint, orderkey_2:bigint]
                                                quantity_0 := "quantity"
                                                suppkey_1 := "suppkey"
                                                orderkey_2 := "orderkey"
                                            - TableScan[local:tpch:lineitem:sf0.01, originalConstraint = null] => [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint]
                                                    orderkey := tpch:orderkey
                                                    partkey := tpch:partkey
                                                    suppkey := tpch:suppkey
                                                    linenumber := tpch:linenumber
                                                    quantity := tpch:quantity
                                                    extendedprice := tpch:extendedprice
                                                    discount := tpch:discount
                                                    tax := tpch:tax
                                                    returnflag := tpch:returnflag
                                                    linestatus := tpch:linestatus
                                                    shipdate := tpch:shipdate
                                                    commitdate := tpch:commitdate
                                                    receiptdate := tpch:receiptdate
                                                    shipinstruct := tpch:shipinstruct
                                                    shipmode := tpch:shipmode
                                                    comment := tpch:comment
                                                    row_number := tpch:row_number

, to pattern:
PlanMatchPattern {
    PlanMatchPattern {
        WindowMatcher{functionCalls=["max"("extendedprice") OVER (PARTITION BY "orderkey" ORDER BY "tax" ASC ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)]}
        PlanMatchPattern {
            PlanMatchPattern {
                WindowMatcher{functionCalls=["avg"("discount") OVER (PARTITION BY "partkey" ORDER BY "receiptdate" ASC ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)]}
                PlanMatchPattern {
                    PlanMatchPattern {
                        WindowMatcher{functionCalls=["sum"("quantity") OVER (PARTITION BY "suppkey" ORDER BY "orderkey" ASC ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)]}
                        anyTree PlanMatchPattern {
                        }
                    }
                }
            }
        }
    }
}
 expected [true] but found [false]
Expected :true
Actual   :false
```

CC: @kokosing

**Comment by kokosing** [comment]
See https://github.com/prestodb/presto/commit/a79c4660ad6aca399b75f5b3b71e352a7fc4d87e.

The problem here is not the verbosity.

The problem is to understand why given comparison fails. It is very similar to REGEX, to find why given pattern does not match to given string. I use debugger and it takes me a while to find the root cause. Additionally that DSL does not support all the plan forms, so it can be the case that some it has to be extended to support above query.

I was working to make it even more verbose, to print all possibilities but this did not help (it was too verbose).

It looks like you are missing `anyTree()` at the beginning in the example above.

**Comment by electrum** [comment]
I had a similar problem with equality bugs in the AST that I solved with the tree linearization step on failure, making it easy to see exactly which part was failing.

I don't know if this is relevant here but it sounds similar.

**Comment by maciejgrzybek** [comment]
@kokosing I completely understand that sometimes DSL framework has to be extended, like @ebd2 did for window functions. The problem is not with this particular query. I used it only as an example to show those who are not familiar with DSL how verbose and not-readable the output is.
If we had error message pointing out where the problem is located instead of failing with boolean match/not match would be much better.
E.g.:
```java
assertUnitPlan("select 1",
                anyTree(
                        tableScan("some_table")));
```
Produces output:
```sql
java.lang.AssertionError: Plan does not match:
 - Output[_col0] => [expr_2:integer]
        _col0 := expr_2
    - Project => [expr_2:integer]
            expr_2 := "expr_1"
        - Project => [expr_1:integer]
                expr_1 := "expr_0"
            - Project => [expr_0:integer]
                    expr_0 := "expr"
                - Project => [expr:integer]
                        expr := 1
                    - Values => []
                            ()

, to pattern:
anyTree PlanMatchPattern {
    PlanMatchPattern {
        TableScanMatcher{expectedTableName=some_table}
    }
}
 expected [true] but found [false]
```
What if we had:
```sql
java.lang.AssertionError: Plan does not match:
 - Output[_col0] => [expr_2:integer]
        _col0 := expr_2
    - Project => [expr_2:integer]
            expr_2 := "expr_1"
        - Project => [expr_1:integer]
                expr_1 := "expr_0"
            - Project => [expr_0:integer]
                    expr_0 := "expr"
                - Project => [expr:integer]
                        expr := 1
                    - Values => []
                            ()

, to pattern:
anyTree PlanMatchPattern {
    PlanMatchPattern {
        TableScanMatcher{expectedTableName=some_table}
    }
}
 expected [TableScan{expectedTableName=some_table}] not found.
```

**Comment by kokosing** [comment]
@maciejgrzybek 
As I said before it is like a REGEX matching. It is hard to tell what is missing `ala ma kta` to match `.*ma ko.*`. Is it just `o` or  `ma ko` at the end... and many other possible solutions.

As I said before, I tried to do something similar to what you suggested, but I ended up with a overly verbose.

I will try with @electrum idea this week and see if that can help.

**Comment by maciejgrzybek** [comment]
@kokosing as per offline discussion: maybe it would be easy to provide such a solution (_expected `node` not found_) for plans without `anyTree` included. Those are easier to follow:
You can traverse both lists (`actual` and `expected`) and compare elements one-by-one reporting first mismatch.
Though, probably most of our DSL tests use `anyTree` anyway...
